File: java/src/network/lobby/packets/user/AvatarUserPacket.java
package network.lobby.packets.user;

import player.property.avatar.Avatar;
import player.property.PlayerProperty;
import misc.Debug;
import misc.compress.*;
import network.lobby.packets.UserPacket;
import network.lobby.packets.PlayerPropertyUserPacket;

public class AvatarUserPacket extends UserPacket implements PlayerPropertyUserPacket
{
	private byte avatarID;

	public AvatarUserPacket(byte id)
	{
		avatarID = id;
	}

	public AvatarUserPacket(AvatarUserPacket packet)
	{
		avatarID = packet.getAvatarID();
	}

	public void assign(AvatarUserPacket avatarPacket)
	{
		Debug.warnIf(avatarPacket == null, "AvatarUserPacket.assign(null)");
		avatarID = avatarPacket.avatarID;
	}

	// Setter
	@Override public void setIDs(byte[] ids)
	{
		avatarID = ids[0];
		if (ids.length != 1)
		{
			Debug.error("AvatarUserPacket.setIDs(): ids enthält mehr als eine id");
		}
	}

	// getter
	public Avatar getAvatar() { return Avatar.getByID(avatarID); }
	public byte getAvatarID() { return avatarID; }
	@Override public PlayerProperty[] getPlayerProperty() { return new PlayerProperty[]{getAvatar()}; }

	public AvatarUserPacket(CompressBuffer buffer)
	{
		avatarID = buffer.cutByte();
	}

	@Override public byte getCID() { return Compressor.AVATAR_USER_PACKET_CID; }
	@Override public byte[] compress()
	{
		return new byte[] {avatarID};
	}
}

File: cpp/src/network/packets/AvatarUserPacket.cpp
#include "AvatarUserPacket.hpp"

#include <misc/Global.hpp>
#include <player/property/avatar/Avatar.hpp>
#include <misc/compress/CompressBuffer.hpp>

AvatarUserPacket::AvatarUserPacket(char c)
        : avatarID(c)
{}

AvatarUserPacket::AvatarUserPacket()
        : avatarID(-1)
{}

AvatarUserPacket::AvatarUserPacket(CompressBuffer* buffer)
{
	avatarID = buffer->cut(1)[0];
}

bool AvatarUserPacket::isValid() const
{
	return avatarID != -1;
}

const std::vector<PlayerProperty*> AvatarUserPacket::getPlayerProperties() const
{
        std::vector<PlayerProperty*> props;
	if (avatarID == -1)
	{
		props.push_back(NULL);
	}
	else
	{
		props.push_back(Avatar::get(avatarID));
	}
        return props;
}

void AvatarUserPacket::setIDs(const std::string& ids)
{
        avatarID = ids[0];
}

std::string AvatarUserPacket::getCompressString() const
{
	std::string s;
	s += avatarID;
	return s;
}

CID AvatarUserPacket::getCID() const
{
        return AVATAR_USER_PACKET_CID;
}

char AvatarUserPacket::getAvatarID() const
{
	return avatarID;
}
